docs/design/2022-08-04-pessimistic-lazy-constraint-check.md

```diff
@@ -0,0 +1,155 @@
+
+A session/global system variable `tidb_constraint_check_in_place_optimistic_only` is introduced to control the behavior. The default value is `ON` and the behavior will remain unchanged. When `tidb_constraint_check_in_place_optimistic_only` is `OFF`, the new behavior will take effect in pessimistic transactions.
+
+```sql
+CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY);
+INSERT INTO t1 VALUES (1), (2);
+
+set tidb_constraint_check_in_place_optimistic_only = off;
+BEGIN PESSIMISTIC;
+SELECT * FROM t1 WHERE id = 1 FOR UPDATE; -- SELECT FOR UPDATE locks key as usual.
+INSERT INTO t1 VALUES (2); -- Skip acquiring the lock and return success.
+COMMIT; -- ERROR 1062 (23000): Duplicate entry '2' for key 'PRIMARY'
+```
+
+#### Protocol
+
+Although we skip locking some keys and doing constraint checks before `COMMIT`, it is still a pessimistic transaction. Except for the keys that need constraint checks, we acquire pessimistic locks for other keys as usual.
+
+It is completely a new kind of behavior to do constraint checks during prewrite for keys that don't get locked ahead of time. We need to change the prewrite protocol to describe it.
+
+Previously, there is a `repeated bool` field describing whether each key needs be locked first. Generally, non-unique index keys don't need to be locked and their conflict checks are skipped.
+
+ ```protobuf
+message PrewriteRequest {
+    // For pessimistic transaction, some mutations don't need to be locked, for example, non-unique index key.
+    repeated bool is_pessimistic_lock = 7;
+}
+ ```
+
+Now, we can reuse this field to describe the keys that need constraint checks during prewrite in a pessimistic transaction:
+
+```protobuf
+message PrewriteRequest {
+	enum PessimisticAction {
+		// The key needn't be locked and no extra write conflict checks are needed.
+		SKIP_PESSIMISTIC_CHECK = 0;
+		// The key should have been locked at the time of prewrite.
+		DO_PESSIMISTIC_CHECK = 1;
+		// The key doesn't need a pessimistic lock. But we need to do data constraint checks.
+		DO_CONSTRAINT_CHECK = 2;
+	}
+    // For pessimistic transaction, some mutations don't need to be locked, for example, non-unique index key.
+    repeated PessimisticAction pessimistic_actions = 7;
+}
+```
+
+Because `enum` is compatible with `bool` on the wire, we can seamlessly extend the  `is_pessimistic_lock` field. In a TiDB cluster, TiKV instances are upgraded before TiDB. So, we can make sure TiKV can handle it correctly when TiDB starts using the new protocol.
+
+When `tidb_constraint_check_in_place_optimistic_only` is off, all the keys that have `PresumeKeyNotExists` flags don't need to be locked when executing the DML. They will be marked with a new special flag `NeedConflictCheckInPrewrite`. When the transaction commits, TiDB will set the actions of these keys to `DO_CONSTRAINT_CHECK`. So, TiKV will not check the existence of pessimistic transactions of these keys. Instead, constraint checks should be done for these keys.
```

Comment:
"Need _conflict_ check" => "Do _constraint_ check" sounds counterintuitive. Would it be better to use the name constraint here, considering that a constraint check implies a conflict check?